### PR TITLE
[FEATURE] Atomic Renderer:  Initial framework 

### PR DESCRIPTION
**Pull request by Shinnnyshinshin**
# New Classes
The PR introduces the Atomic Renderer, a rendering framework that will be shared by both OSS and GE Cloud

The PR introduces the components needed for the first (and most common) type : `StringValueType`, with `TableType` and `ChartType` left as TODOs 

At the top level we introduce `RenderedAtomicContent`:
    - This is the object that is actually sent back to Mercury from OSS GE. 
    - It contains the following fields: 
        1. name : string that describes the renderer that the object is coming from (ie. `atomic.prescriptive.summary`)
        2. value : `RenderedAtomicValue` (a new class introduced in this PR, described in more detail below)
        3. valuetype : string that can describes `RenderedAtomicValue`. Can be `"StringValueType", or "TableType"` etc

`RenderedAtomicValue` is the object that contains the rendered content that will be transformed into its corresponding Mercury model.   Instead of having a separate class for `StringValueType`, `TableType` and `GraphType`, they are all handled together under `RenderedAtomicValue`, and the `valuetype`
string describes which model that Mercury needs to use.  Defining it in this way allows for basic a schema to be defined at the OSS level, and for the actual validation to happen at the Mercury-level. 

# New Methods

1. `_atomic_prescriptive_template()`
  - contains the core logic for generating Prescriptive rendered content, and is shared by: 
    - `_atomic_prescriptive_summary()` --> Mercury
    - `_prescriptive_renderer()` --> DataDocs
    
  - this function is defined in the base `Expectation` class and is called if the Expectation does not have its own `_atomic_prescriptive_template()` function defined. 

2. `_atomic_prescriptive_summary()`

    - called by Mercury. Will call the `_atomic_prescriptive_template()` defined in the specific Expectation, or the one that is in the base Expectation class. 

3. `_prescriptive_renderer()`

    - called by OSS GE in generating DataDocs. Has been modified to call `_atomic_prescriptive_template()` function

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
- CLOUD-250

Definition of Done
- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/render/types/__init__.py [hunk 1]
```diff
@@ -493,3 +494,15 @@ def to_json_dict(self):
...
         d["section_name"] = self.section_name
         return d
+
+
+class RenderedAtomicContent(RenderedContent):
+    def __init__(
+        self,
+        name: str = None,
+        value: RenderedAtomicValue = None,
+        valuetype: str = None,
```
Comment (cdkini) [inline great_expectations/render/types/__init__.py]: Would you mind updating these type annotations to make all these `Optional` (due to the default `None` assignment)?

File: great_expectations/data_context/types/base.py [hunk 2]
```diff
@@ -2262,6 +2262,50 @@ class CheckpointValidationConfigSchema(Schema):
...
+
+class RenderedAtomicValue(DictDot):
+    def __init__(
+        self,
+        template: str = None,
+        params: dict = None,
+        schema: dict = None,
+        # header: list = None,
+        # header_row: str = None,
+        # table: list = None,
```
Comment (cdkini) [inline great_expectations/data_context/types/base.py]: Could we remove the unneeded fields and make all these types `Optional`?

File: great_expectations/expectations/expectation.py [hunk 3]
```diff
@@ -163,6 +169,56 @@ def _validate(
...
 
+    @classmethod
+    def _atomic_prescriptive_template(
+        cls,
+        configuration=None,
+        result=None,
+        language=None,
+        runtime_configuration=None,
+        **kwargs,
+    ):
```
Comment (cdkini) [inline great_expectations/expectations/expectation.py:180]: Could we possibly add a docstring here and below? I usually don't look for them in private methods but this is a relatively complex concept so a general explanation at the parent class level could be helpful!

Doesn't need to be complex

File: great_expectations/expectations/expectation.py [hunk 4]
```diff
@@ -163,6 +169,56 @@ def _validate(
...
+        runtime_configuration = runtime_configuration or {}
```
Comment (cdkini) [inline great_expectations/expectations/expectation.py]: Could we possibly use the following instead?

```
if runtime_configuration is None:
    runtime_configuration = {}
```

I think what you've used makes perfect sense but this tends to be more of a JavaScript style of conditional. Minor but might help future users 😄

File: great_expectations/expectations/expectation.py [hunk 5]
```diff
@@ -454,15 +510,7 @@ def _diagnostic_unexpected_table_renderer(
...
-    @renderer(renderer_type="renderer.diagnostic.observed_value")
-    def _diagnostic_observed_value_renderer(
-        cls,
-        configuration=None,
-        result=None,
-        language=None,
-        runtime_configuration=None,
-        **kwargs,
-    ):
+    def _get_observed_value_from_evr(self, result: ExpectationValidationResult):
```
Comment (cdkini) [inline great_expectations/expectations/expectation.py]: If we're typing the args could we also type the return value of `str`?

File: great_expectations/expectations/expectation.py [hunk 6]
```diff
@@ -482,6 +530,43 @@ def _diagnostic_observed_value_renderer(
         else:
             return "--"
 
+    @classmethod
+    @renderer(renderer_type="atomic.diagnostic.observed_value")
+    def _atomic_diagnostic_observed_value(
```
Comment (cdkini) [inline great_expectations/expectations/expectation.py:544]: Same general comment about docstrings here. I don't have a great deal of familiarity with all these renderers so if you think these are easily understandable, please ignore my comments regarding documentation.

File: great_expectations/expectations/registry.py [hunk 7]
```diff
@@ -63,10 +63,26 @@ def register_renderer(
...
 
 
+def get_renderer_names(object_name: str) -> List[str]:
+    return list(_registered_renderers.get(object_name, {}).keys())
+
+
+def get_renderer_impls(object_name: str) -> list:
+    return list(_registered_renderers.get(object_name, {}).values())
+
+
```
Comment (cdkini) [inline great_expectations/expectations/registry.py]: Is this duplicated? Could we keep the first set of functions since they've got more explicit typing?

**Review by cdkini — APPROVED** [review]
I don't have a great grasp on what exactly these renderer functions are doing since I haven't been a part of the design process but this all looks solid from a Python perspective!

I'll sign off here but maybe get another set of eyes on this if you want a more thorough review.

File: great_expectations/data_context/types/base.py (hunk 2 above)
Comment (Shinnnyshinshin) [reply great_expectations/data_context/types/base.py]: header, header_row and table are placeholders for when TableType is implemented. Is It ok to keep those? I'll add the Optional

File: great_expectations/expectations/expectation.py (hunk 4 above)
Comment (Shinnnyshinshin) [reply great_expectations/expectations/expectation.py]: Sure thing

File: great_expectations/expectations/registry.py (hunk 7 above)
Comment (Shinnnyshinshin) [reply great_expectations/expectations/registry.py]: It is 😓  Thanks for catching this.

File: great_expectations/render/types/__init__.py (hunk 1 above)
Comment (Shinnnyshinshin) [reply great_expectations/render/types/__init__.py]: done

**Pull request merged**